### ISSUE TITLE
Mix.Task.run("archive.install") removes projects paths from the code path

### ISSUE THREAD

**Issue by virtual-light** [elixir-lang/elixir#12784]
### Elixir and Erlang/OTP versions

```
Erlang/OTP 26 [erts-14.0.1] [source] [64-bit] [smp:20:20] [ds:20:20:10] [async-threads:1] [jit:ns]

Elixir 1.15.0 (compiled with Erlang/OTP 26)

```

### Operating system

5.15.0-76-generic GNU/Linux (Ubuntu 22.04.2 LTS)

### Current behavior

Example: 

**test_project/lib/mix/tasks/mtest.ex**

```elixir
defmodule Mix.Tasks.Mtest do
  use Mix.Task

  @impl Mix.Task
  def run(_args) do
    dbg(:code.lib_dir(:test_project))
    
    Mix.Task.run("local.hex", ["--force"])

    dbg(:code.lib_dir(:test_project))

    Mix.Task.run("archive.install", ["hex", "phx_new", "~> 1.7.7", "--force"])
    
    dbg(:code.lib_dir(:test_project))
  end
end
```

```
mix mtest
```

Output:
```bash
[lib/mix/tasks/mtest.ex:6: Mix.Tasks.Mtest.run/1]
:code.lib_dir(:test_project) #=> ~c"/home/test/Tests/test_project/_build/dev/lib/test_project"

* creating /home/test/.asdf/installs/elixir/1.15.0-otp-26/.mix/archives/hex-2.0.6

[lib/mix/tasks/mtest.ex:10: Mix.Tasks.Mtest.run/1]
:code.lib_dir(:test_project) #=> ~c"/home/test/Tests/test_project/_build/dev/lib/test_project"

* creating /home/test/.asdf/installs/elixir/1.15.0-otp-26/.mix/archives/phx_new-1.7.7

[lib/mix/tasks/mtest.ex:14: Mix.Tasks.Mtest.run/1]
:code.lib_dir(:test_project) #=> {:error, :bad_name}
```

`Mix.Task.run("archive.install")` removes all paths for loaded projects and their dependencies from code path. 

### Expected behavior

`Mix.Task.run("archive.install")` should preserve code paths that had been added before command run. 

Current workaround is to get existing projects paths and add them again with `:code.add_path/2` or `:code.set_path/2` after command run.

**Comment by josevalim** [comment]
Every time a project is compiled, code paths are removed. Is there a reason why you want them to be preserved?

**Comment by virtual-light** [comment]
@josevalim 

I thought compilation happens inside `Mix.Local.Installer.MixProject` and `Phx.New.MixProject` but not in running project. 

It is confusing that after the command run current project paths have been removed and it requires to add them again to be able to continue if you have additional code to run after. I expected some kind of isolation for current project from running `Mix.Task.run("archive.install")`.  For example `Mix.Task.run("local.hex", ["--force"])` doesn't removes paths.

**Comment by josevalim** [comment]
We don't shell out, everything happens in the same instance. local.hex does not compile anything. I still need to understand your use case better, is there a reason why you want them to be preserved?

**Comment by virtual-light** [comment]
Debug says its compiles in different projects:

```
Running mix will_recompile (inside Mix.Local.Installer.MixProject)
...
Running mix compile (inside Phx.New.MixProject)
```
And why current project should even be recompiled?`archive.install` installs not a project's but a global dependency after all.

My case is an task that is an extension for `phx.new`. Firstly it creates project with `phx.new` and then does some additional custom bootstrapping. I ran into issue that after installing `phx_new` things like `Application.app_dir/1` and other related to code paths started to say that no files available.

**Comment by virtual-light** [comment]
Unfortunately `Mix.ensure_application!(:your_extension)` doesn't work.  

```
** (Mix.Error) The application "test_project" could not be found. This may happen if your Operating System broke Erlang into multiple packages and may be fixed by installing the missing "erlang-dev" and "erlang-test_project" packages

```

@josevalim 
Also wanted to mention that on old versions of elixir/opt (`1.12-otp-24`) and phx_new (1.5.3)  behavior has been different.  

Mix.Task.run("archive.install") haven't removed paths.

**Comment by josevalim** [comment]
It should work on Elixir v1.5.2. In any case, code path pruning is a new feature to improve compile times and make applications closer to prod. If you do the code.set_path trick, it should work on all cases. Thanks!

**Comment by virtual-light** [comment]
FYI: Recompilation also removes project's modules structures from consolidated protocols.
Thats also confusing because recompilation doesn't unload modules that holding those structures.

@josevalim

**Comment by virtual-light** [comment]
So in the beginning `String.Chars` haven't been loaded
```elixir
:code.is_loaded(String.Chars) #=> false
```  

but available  
```elixir
:code.which(String.Chars) #=> ~c"/home/test/Tests/test_project/_build/prod/lib/test_project/consolidated/Elixir.String.Chars.beam"
```

And had consolidated protocol with custom type impl

```elixir
String.Chars.__protocol__(:impls) #=> {:consolidated,
 [Atom, BitString, Date, DateTime, Float, Integer, List, My.Module,
  NaiveDateTime, Time, URI, Version, Version.Requirement]}
```

After 
```elixir
Mix.Task.run("archive.install", ["hex", "phx_new", "~> 1.7.7", "--force"])
```

`String.Chars` became substituted by :phx_new

```elixir
:code.is_loaded(String.Chars) #=> {:file,
 ~c"/tmp/mix-local-installer-fetcher-Ardjqg/deps/phx_new/_build/prod/lib/phx_new/consolidated/Elixir.String.Chars.beam"}
```

```elixir
:code.which(String.Chars) #=> ~c"/tmp/mix-local-installer-fetcher-Ardjqg/deps/phx_new/_build/prod/lib/phx_new/consolidated/Elixir.String.Chars.beam"
```


```elixir
String.Chars.__protocol__(:impls) #=> {:consolidated,
 [Atom, BitString, Date, DateTime, Float, Integer, List, NaiveDateTime, Time,
  URI, Version, Version.Requirement]}
```


Recompilation affects current project state hard leaving artifacts of last compilation. 
So to continue it requires not only manually add code paths but also load original modules back.

**Comment by josevalim** [comment]
I think we can disable protocol consolidation on archive.install. I don't see any reason to perform it.

**Comment by josevalim** [comment]
Can you please give the v1.15 branch a try?

**Comment by virtual-light** [comment]
Got same behavior with modules substitutions / protocols impls on  `Elixir (1.15.0)`.


Btw found strange httpc leak with recompilation:

Normal case
```sh
Erlang/OTP 26 [erts-14.0.1] [source] [64-bit] [smp:20:20] [ds:20:20:10] [async-threads:1] [jit:ns]

Interactive Elixir (1.15.3) - press Ctrl+C to exit (type h() ENTER for help)
iex(1)> :httpc.request("https://google.com")
{:ok,
# ...
``` 

Recompilation:

```sh
Erlang/OTP 26 [erts-14.0.1] [source] [64-bit] [smp:20:20] [ds:20:20:10] [async-threads:1] [jit:ns]

Interactive Elixir (1.15.3) - press Ctrl+C to exit (type h() ENTER for help)
iex(1)> Mix.Task.run("local.hex", ["--force"])
* creating /home/test/.asdf/installs/elixir/1.15.3-otp-26/.mix/archives/hex-2.0.6
true
iex(2)> Mix.Task.run("archive.install", ["hex", "phx_new", "--force"])
Resolving Hex dependencies...
Resolution completed in 0.022s
New:
  phx_new 1.7.7
* Getting phx_new (Hex package)
All dependencies are up to date
Compiling 11 files (.ex)
Generated phx_new app
Generated archive "phx_new-1.7.7.ez" with MIX_ENV=prod
* creating /home/test/.asdf/installs/elixir/1.15.3-otp-26/.mix/archives/phx_new-1.7.7
true
iex(3)> :httpc.request("https://google.com")
{:error,
 {:shutdown,
  {{:error, :undef},
   [
     {:http_chunk, :decode,
      ["4e8b\r\n<!doctype html><html itemscope=\"\" itemtype=\"http://schema.org/WebPage\" lang=\"uk\"><head><meta content=\"text/html; charset=UTF-8\" http-equiv=\"Content-Type\"><meta co",
```

However if to make a `:httpc.request` before recompilation then error is missing:

```sh
Erlang/OTP 26 [erts-14.0.1] [source] [64-bit] [smp:20:20] [ds:20:20:10] [async-threads:1] [jit:ns]

Interactive Elixir (1.15.3) - press Ctrl+C to exit (type h() ENTER for help)
iex(1)>  :httpc.request("https://google.com")
{:ok,
# ...

iex(2)> Mix.Task.run("local.hex", ["--force"])
* creating /home/test/.asdf/installs/elixir/1.15.3-otp-26/.mix/archives/hex-2.0.6
true
iex(3)> Mix.Task.run("archive.install", ["hex", "phx_new", "--force"])
Resolving Hex dependencies...
Resolution completed in 0.021s
New:
  phx_new 1.7.7
* Getting phx_new (Hex package)
All dependencies are up to date
Compiling 11 files (.ex)
Generated phx_new app
Generated archive "phx_new-1.7.7.ez" with MIX_ENV=prod
* creating /home/test/.asdf/installs/elixir/1.15.3-otp-26/.mix/archives/phx_new-1.7.7
true
iex(4)> :httpc.request("https://google.com")
{:ok,
```
Don't know from where `4e8b\r\n` comes from but current process message box is empty. 
Same behavior with Elixir 1.15.0

**Comment by josevalim** [comment]
I don't mean v1.15.0, I meant v1.15 which means compiling from source on GitHub. :)

**Comment by virtual-light** [comment]
@josevalim 

Tested on v1.15 branch. Can confirm that protocol issue because of module substitution have been fixed there :+1: 


However I still struggle because of this `4e8b\r\n` httpc` issue I described in the previous post.  It clearly comes from `:ssl` because `http://` requests work fine. And it happens only after recompilation. Interesting that if you first do at least one request before recompilation then it works fine after recompilation but if not - issue remains for all further request. 

What can be an issue and possible workaround for it? Should I create separate issue to for it?

**Comment by josevalim** [comment]
It may be that something has already configured httpc. You should probably stop the inets application and start it again. But you will always be battling side-effects from other applications. In my suggestion you should shell out for the `mix archive.install`.

**Comment by virtual-light** [comment]
Makes sense. Shelling out makes it slower and also requires to shell out all dependent commands (for example, if shell out `mix archive.install hex phx_new --force` you can't run `Mix.Task.run("phx.new", ["new_project"])` in running shell because it doesn't see `phx.new` task) which makes it even more slower. 
However in case of a tool it is affordable price to pay for desired isolation.  

Thank you for your assistance @josevalim
